### PR TITLE
PDF text selection & highlights: various fixes

### LINKED ISSUE #7928: No spaces in highlights with Greek text.
* KOReader version: 2021/06
* Device: PocketBook Touch Lux 5 (PB 628)

#### Issue
The text of the highlights containing Greek text has no spaces between the words. Could this be fixed, please?

#### Steps to reproduce
When I highlight a Greek text, even in an epub file, this text, in highlight display, hasn't any space between the words.

Thank you!

### PR DESCRIPTION
**Pull request by poire-z**
`PDF text selection: fix/tweak spacing between words/boxes`

We may get multiple boxes when selecting texts, one for each word, and we have to add spaces between the extracted words ourselves. Previously, we were only adding a space if the last char of previous word was ASCII, so missing spaces after accents or greek words.
Try to do better by measuring the distances between boxes and comparing to box heights, with a few heuristics.

`PDF written highlights: fix boxes, trash cached tiles`

TileCacheItem: add created_ts property.
Document: manage a tile_cache_validity_ts and ignore older cached tiles.
This timestamps is updated when highlights are written as annotations in, or deleted from, the PDF, so we can get the most current rendered bitmap from MuPDF and avoid highlight ghosts on old tiles.
Save this timestamp in doc settings so older cached to disk tiles will also be ignored across re-openings.
This commit will be added a bump for https://github.com/koreader/koreader-base/pull/1388.

See https://github.com/koreader/koreader/issues/7928#issuecomment-882089306.
Hopefully close #7928.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/7988)

### CODE REVIEW

File: frontend/document/document.lua [hunk 1]
```diff
@@ -386,7 +398,16 @@ function Document:renderPage(pageno, rect, zoom, rotation, gamma, render_mode)
         hash_excerpt = hash.."|"..tostring(rect)
         tile = DocCache:check(hash_excerpt)
     end
-    if tile then return tile end
+    if tile then
+        if self.tile_cache_validity_ts then
+            if tile.created_ts and tile.created_ts >= self.tile_cache_validity_ts then
+                return tile
+            end
+            logger.dbg("discarding staled cached tile")
```
Comment (NiLuJe) [inline frontend/document/document.lua:406]:
```suggestion
            logger.dbg("discarding stale cached tile")
```

**Comment by NiLuJe** [comment]
The tile cache logic looks neater that the original PoC, :+1: ;).

**Comment by poire-z** [comment]
~~I guess/hope it was already this for before this PR - but it feels like this save highlight stuff is quite a mess.
Tested on my Android phone, with "Write highlights> prompt":
highlighting some text gets it in yellow+gray (so, written into the PDF)
switching documents, I make sure to always hit "no save".
back to this document, still in yellow+gray
hit delete
switching documents, I make sure to always hit "no save".
back to this document, still in yellow - and as no gray, no ability to remove as its now written in the PDF...~~

No, that is a tile cache issue. Looks like my tile cache validation stuff has issues on Android :/
(But it's still confusing with "prompt": things happen, are written in memory in the pdf, are shown on screen, and you get the prompt at the end which somehow just "erase" all that was happening :)

**Comment by poire-z** [comment]
> No, that is a tile cache issue. Looks like my tile cache validation stuff has issues on Android :/

No, not specific to Android, I get these confusing behaviour on the emulator too.
Highlight > save > reload > tap to unhighlight > don't save, and you get a sticky yellow unremovable highlight - which is somehow normal with these stupid steps.

Still a small tile cache issue: after tap to unhighlight, the tile bitmap has no highlight - but on reload, we get this not-unvalidated cached tile withouth the highlight - but the highlight is still in the PDF, and should be shown in yellow (and is if you change zoom mode).
I guess at save/don't save time, we should also invalidate the cached tiles, whatever the choice.

**Comment by Frenzie** [comment]
Sounds like something else should be added to the hash, though I couldn't say what exactly.

**Comment by NiLuJe** [comment]
> Sounds like something else should be added to the hash, though I couldn't say what exactly.

Yeah, I was thinking the *amount* of highlights on the page might be the least hairy approach, but it's obviously not super accurate.

**Comment by poire-z** [comment]
We shouldn't need to add anything to the hash.
My issue is probably that I should reset the cache validity timestamps in a few more places (like when we hit "don't save" when prompted - as current tile is not identical to the one we'll get on next load if it were not cached).

**Comment by poire-z** [comment]
Small #8002 should avoid the above situations.

**Pull request merged**